src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractExtensionMethodImportCompletionProvider.cs

```diff
@@ -0,0 +1,95 @@
+
+        protected async override Task AddCompletionItemsAsync(
+            CompletionContext completionContext,
+            SyntaxContext syntaxContext,
+            HashSet<string> namespaceInScope,
+            bool isExpandedCompletion,
+            CancellationToken cancellationToken)
+        {
+            using (Logger.LogBlock(FunctionId.Completion_ExtensionMethodImportCompletionProvider_GetCompletionItemsAsync, cancellationToken))
+            {
+                var syntaxFacts = completionContext.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+                if (TryGetReceiverTypeSymbol(syntaxContext, syntaxFacts, out var receiverTypeSymbol))
+                {
+                    var items = await ExtensionMethodImportCompletionHelper.GetUnimportedExtensionMethodsAsync(
+                        completionContext.Document,
+                        completionContext.Position,
+                        receiverTypeSymbol,
+                        namespaceInScope,
+                        forceIndexCreation: isExpandedCompletion,
+                        cancellationToken).ConfigureAwait(false);
+
+                    completionContext.AddItems(items.Select(i => Convert(i)));
+                }
+            }
+        }
+
+        private static bool TryGetReceiverTypeSymbol(SyntaxContext syntaxContext, ISyntaxFactsService syntaxFacts, [NotNullWhen(true)] out ITypeSymbol? receiverTypeSymbol)
+        {
+            var parentNode = syntaxContext.TargetToken.Parent;
+
+            // Even though implicit access to extension method is allowed, we decide not support it for simplicity 
+            // e.g. we will not provide completion for unimport extension method in this case
+            // New Bar() {.X = .$$ }
+            var expressionNode = syntaxFacts.GetLeftSideOfDot(parentNode, allowImplicitTarget: false);
+
+            if (expressionNode == null)
+            {
+                receiverTypeSymbol = null;
+                return false;
+            }
+
+            receiverTypeSymbol = syntaxContext.SemanticModel.GetTypeInfo(expressionNode).Type;
+
+            if (receiverTypeSymbol is IErrorTypeSymbol errorTypeSymbol)
+            {
+                receiverTypeSymbol = errorTypeSymbol.CandidateSymbols.Select(s => GetSymbolType(s)).FirstOrDefault(s => s != null);
+            }
+
+            return receiverTypeSymbol != null;
+        }
+
+        private static ITypeSymbol? GetSymbolType(ISymbol symbol)
+            => symbol switch
+            {
+                ILocalSymbol localSymbol => localSymbol.Type,
+                IFieldSymbol fieldSymbol => fieldSymbol.Type,
+                IPropertySymbol propertySymbol => propertySymbol.Type,
+                IParameterSymbol parameterSymbol => parameterSymbol.Type,
+                IAliasSymbol aliasSymbol => aliasSymbol.Target as ITypeSymbol,
+                _ => symbol as ITypeSymbol,
+            };
```

Comment:
this seems weird.  extension methods are not called off of symbols.  They're explicitly called off of type-instances.  So we need to just do GetTypeInfo(...) and then validate that GetSymbolInfo(...) does **not** return an ITypeSymbol.

Reply:
Hmm, That's exactly what I did earlier, but I have changed it to this after your comment, I probably didn't think clearly before doing so. o_o
https://github.com/dotnet/roslyn/blob/e2f0c425bfa3c767ecb01ba3483e54e16a0ae6e2/src/Features/CSharp/Portable/Completion/CompletionProviders/ImportCompletion/ExtensionMethodImportCompletionProvider.cs#L38

Reply:
indeed. i was either unclear or mistaken.  :)

i think we can simplify this though into something like this:

```
bool ShouldShowExtensionMethodCompletion(ExpressionSyntax expr)
{
    // add docs
    if (sm.GetSymbolInfo(expr, ct).GetAnySymbol() is ITypeSymbol)
        return false;

    return sm.GetTypeInfo(expr, ct).Type?.TypeKind != TypeKind.ErrorType;
}
```

That's simple and clear and can be doc'ed as to why each clause exists.  We don't have extension methods exposed off of types, and the expression we're calling off of needs to have an actual instance type.

Reply:
Sounds good, just needs to be a little more tolerant to errors here to handle cases like this

```cs
x.$$
var z = 10;
```